src/Encode/AlphaEncodingManager.cpp

```diff
@@ -1,111 +1,238 @@
+    {"G", VRCommDataAlphaEncodingKey::JoyY},      // joystick y component
+    {"H", VRCommDataAlphaEncodingKey::JoyBtn},    // joystick button
+    {"I", VRCommDataAlphaEncodingKey::BtnTrg},    // trigger button
+    {"J", VRCommDataAlphaEncodingKey::BtnA},      // A button
+    {"K", VRCommDataAlphaEncodingKey::BtnB},      // B button
+    {"L", VRCommDataAlphaEncodingKey::GesGrab},   // grab gesture (boolean)
+    {"M", VRCommDataAlphaEncodingKey::GesPinch},  // pinch gesture (boolean)
+    {"N", VRCommDataAlphaEncodingKey::BtnMenu},   // system button pressed (opens SteamVR menu)
+    {"O", VRCommDataAlphaEncodingKey::BtnCalib},  // calibration button
+    {"", VRCommDataAlphaEncodingKey::Null},       // Junk key
 };
 
-static std::string getArgumentSubstring(const std::string& str, const char del) {
-  const size_t start = str.find(del);
+static const std::map<VRCommDataAlphaEncodingKey, std::string> VRCommDataAlphaEncodingOutputKeyString{
+    {VRCommDataAlphaEncodingKey::FinThumb, "A"},   // thumb force feedback
+    {VRCommDataAlphaEncodingKey::FinIndex, "B"},   // index force feedback
+    {VRCommDataAlphaEncodingKey::FinMiddle, "C"},  // middle force feedback
+    {VRCommDataAlphaEncodingKey::FinRing, "D"},    // ring force feedback
+    {VRCommDataAlphaEncodingKey::FinPinky, "E"}    // pinky force feedback
+};
 
-  if (start == std::string::npos) return std::string();
+static std::map<VRCommDataAlphaEncodingKey, std::string> ParseInputToMap(const std::string& str) {
+  std::map<VRCommDataAlphaEncodingKey, std::string> result;
 
-  const size_t end =
-      str.find_first_of(VRCommDataAlphaEncodingCharacters, start + 1);  // characters may not necessarily be in order, so end at any letter
+  int i = 0;
+  while (i < str.length()) {
+    // Advance until we get an alphabetic character (no point in looking at values that don't have a key associated with them)
 
-  return str.substr(start + 1, end - (start + 1));
-}
+    if (str[i] >= 0 && str[i] <= 255 && IsCharacterKeyCharacter(str[i])) {
+      std::string key = {str[i]};
+      i++;
 
-static bool argValid(const std::string& str, const char del) {
-  return str.find(del) != std::string::npos;
-}
+      // we're going to be parsing a "long key", i.e. (AB) for thumb finger splay. Long keys must always be enclosed in brackets
+      if (key[0] == '(') {
+        while (str[i] >= 0 && str[i] <= 255 && IsCharacterKeyCharacter(str[i]) && i < str.length()) {
+          key += str[i];
+          i++;
+        }
+      }
+
+      std::string value = "";
+      while (str[i] >= 0 && str[i] <= 255 && isdigit(str[i]) && i < str.length()) {
+        value += str[i];
+        i++;
+      }
 
-AlphaEncodingManager::AlphaEncodingManager(const float maxAnalogValue) : EncodingManager(maxAnalogValue) {}
+      // Even if the value is empty we still want to use the key, it means that we have a button that is pressed (it only appears in the packet if it
```

Comment:
Is it safe to make this assumption or would it be a good idea to have a contingency for value-dependent fields?

Reply:
I think we have to, buttons at the moment don't have a value associated with them. Ideally each input would specify a default value (or a value representing something that shouldn't be parsed) when there is no value associated with it and it expects one.

Reply:
So in this case, if a input was passed without any value, does it zero out or does it keep the last valid value that it has?

Reply:
if it's not provided a value it'll go to what the default is, I think we need to keep this functionality as currently buttons don't have a value associated with them though.

Reply:
Alright, no problem. Later on we might want to add an "input memory" kind of functionality so that when invalid inputs are read (from comms noise etc) it just uses the last valid value. But for now, this is definitely fine.